### PR TITLE
refactor: add logical plan recursive equality testing

### PR DESCRIPTION
**Pull request by bcallender**
# refactor: add logical plan recursive equality testing

This PR adds a comprehensive implementation of `__eq__` for logical plans to enable proper equality testing. The implementation:

- Adds a base `__eq__` method to `LogicalPlan` that handles type checking and recursive comparison
- Implements `_eq_specific` methods for all plan types to compare their specific attributes
- Fixes a typo in the base expression equality method (`self_chldren` → `self_children`)
- Adds equality support for `Schema` and `ColumnField` classes
- Includes test coverage for all plan types and edge cases
- Simplifies the serialization tests to use the new equality methods instead of string representation comparisons

### CODE REVIEW

**Comment by bcallender** [comment]
* **#109** <a href="https://app.graphite.dev/github/pr/typedef-ai/fenic/109?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#108** <a href="https://app.graphite.dev/github/pr/typedef-ai/fenic/108?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#107** <a href="https://app.graphite.dev/github/pr/typedef-ai/fenic/107?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#133** <a href="https://app.graphite.dev/github/pr/typedef-ai/fenic/133?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/typedef-ai/fenic/133?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/fenic/core/types/schema.py [hunk 1]
```diff
@@ -120,3 +137,18 @@ def column_names(self) -> List[str]:
...
+            other: The other schema to compare to.
+
+        Returns:
+            bool: True if the schema has equal non schema attributes to the other schema, False otherwise.
+        """
+        if not isinstance(other, Schema):
+            return False
+        if len(self.column_fields) != len(other.column_fields):
+            return False
+        return all(field1 == field2 for field1, field2 in zip(self.column_fields, other.column_fields, strict=False))
```
Comment (bcallender) [inline src/fenic/core/types/schema.py]: Review Note: Does order matter here? If I have a schema with the exact same columns, but in a different order, should that count?

File: src/fenic/core/_logical_plan/plans/base.py [hunk 2]
```diff
@@ -118,3 +118,30 @@ def with_children(self, children: List[LogicalPlan], session_state: Optional[Bas
...
+            return False
+        if not self._eq_specific(other):
+            return False
+        self_children = self.children()
+        other_children = other.children()
+        if len(self_children) != len(other_children):
+            return False
+        return all(
+            child1 == child2
+            for child1, child2 in zip(self_children, other_children, strict=False)
```
Comment (rohitrastogi) [inline src/fenic/core/_logical_plan/plans/base.py]: nit: we should do strict=True to be defensive (here and otherwise).

File: src/fenic/core/_logical_plan/plans/transform.py [hunk 3]
```diff
@@ -579,12 +621,52 @@ def __init__(
...
+        return SemanticCluster(
+            input,
+            by_expr,
+            num_clusters,
+            max_iter,
+            num_init,
+            label_column,
+            centroid_column,
+            session_state,
+            None,
```
Comment (rohitrastogi) [inline src/fenic/core/_logical_plan/plans/transform.py]: while we are here, its clearer to just pass session_state as a named arg and ignore the schema field

**Review by rohitrastogi — APPROVED** [review]
